prow/gerrit/adapter/adapter.go

```diff
@@ -168,6 +169,43 @@ func createRefs(reviewHost string, change client.ChangeInfo, cloneURI *url.URL,
 	return refs, nil
 }
 
+// failedJobs find jobs currently reported as failing (used for retesting).
+//
+// Failing means the job is complete and not passing.
+// Scans messages for prow reports, which lists jobs and whether they passed.
+// Job is included in the set if the latest report has it failing.
+func failedJobs(account int, revision int, messages ...gerrit.ChangeMessageInfo) sets.String {
+	failures := sets.String{}
+	times := map[string]time.Time{}
+	for _, message := range messages {
+		if message.Author.AccountID != account { // Ignore reports from other accounts
+			continue
+		}
+		if message.RevisionNumber != revision { // Ignore reports for old commits
+			continue
+		}
+		// TODO(fejta): parse triggered job reports and remove from failure set.
+		// (alternatively refactor this whole process rely less on fragile string parsing)
+		report := reporter.ParseReport(message.Message)
+		if report == nil {
+			continue
+		}
+		for _, job := range report.Jobs {
+			name := job.Name
+			if latest, present := times[name]; present && message.Date.Before(latest) {
+				continue
+			}
+			times[name] = message.Date.Time
```

Comment:
Update times[name] for every message could cause out of order problem, unless messages are already sorted by datetime. 
For example, a job passed then failed, but the messages arrives in reversed order, then this test is not considered failure

Reply:
That scenario is what this code guards against.

If we have already seen this job in a report from a later message, then we will skip this job for the current report. 

`TestFailedJobs` unit tests this scenario (see the redshifted scenario)

Reply:
I see, didn't notice the `continue` above

Not sure I get the idea of redshifted analog though :)

Reply:
Gravity slows down time. This is one (improbable) reason why a message with an earlier timestamp arrives after a message with a later timestamp

https://www.sciencealert.com/earth-s-core-is-2-5-years-younger-than-its-crust-thanks-to-the-curvature-of-space-time

https://www.youtube.com/watch?v=EXHebyV2TNM